### ISSUE TITLE
[FEATURE] Elastic 8 Client - Stores Enums as Strings

### ISSUE THREAD

**Issue by DR9885** [elastic/elasticsearch-net#8194]
**Is your feature request related to a problem? Please describe.**
Before enums were stored as integers, which allowed us to do range queries. 

Ex: we have a Rating Enum, and sometimes we like to search above a certain rating.

**Describe the solution you'd like**
Options:
1) Go back to enums defaulting to integers.
2) Add a way to configure an enum field as integer or string.

**Comment by DR9885** [comment]
Good to know, we can leverage the System.Text features.

However, these changes have other impacts:
- Range Queries are not supported. (as stated above)
- For Everyone, Upgrading from Nest to 8, All Term Queries will no longer work, and will have to go to ".keyword"
- Text & Keyword is much larger than Int, so when having a large search db with a bunch of enums. This will now increase the size and slow down the index speed.
- Responses will also be slower, with larger text to be sent over the network. some enum names can be really long.
- Also, C# has flags, which are designed to store attributes in the smallest format, which would break the purpose of this.
https://dev.to/ayodejii/enum-flags-in-c-20a6
- most nosql stores keep enums as int format. 


Example of some Enum Values we use, that would be too wordy to store as strings.
```
        SuccessorAdditionalOrChangeInTrustee = 52,
        UnscheduledDrawCreditEnhancement = 53,
        SubstitutionCreditLiquidityProvider = 54 ,
        FinancialObligationIncurrence = 55,
        FinObligEventReflectingFinancialDifficulties = 56 , 
        FinancialObligationIncurrenceDebtObligation  = 57 ,
        FinancialObligIncurrenceGuaranteeDebtOblig = 58 ,
        FinanObligIncurrenceGuaranteeDerivativeInstr = 59,
        FinancialObligIncurrenceDerivativeInstrument = 60,
        FinanObligEventReflectFinDifficultiesDefault = 61,
        FinObligEventReflectFinDiffEventOfAccel = 62,
        FinObligEventReflectFinDiffTerminationEvent = 63,
        FinanObligEventReflectFinDiffModifOfTerms = 64 ,
        FinanObligEventReflectFinanDiffOther = 65,
        InitialAssetBackedSecuritiesDisclosure = 66 ,
        QuarterlyAssetBackedSecuritiesDisclosure = 67 ,
        AnnualAssetBackedSecuritiesDisclosure = 68,
        OtherAssetBackedSecuritiesDisclosure = 69 ,
        BankLoanAlternativeFinancingFilings = 70 
```

**Comment by flobernd** [comment]
The new default behavior was decided before I joined the team and there probably are as well good arguments in favor of this change. Maybe @Mpdreamz @stevejgordon can remember these?

> Range Queries are not supported. (as stated above)

Range queries on enum fields is a nieche use-case and probably even indicates a wrong type. Enums are not intended to cover a continuous range of values.

> For Everyone, Upgrading from Nest to 8, All Term Queries will no longer work, and will have to go to ".keyword"

Term query is the wrong thing to use for enum values anyways. A match query would be more approriate imo, because you are looking for an exact match and not a semantic term.

However, your point is completely valid. This is a breaking change and could mess with existing data, if not carefully handled (like e.g. using the workaround I mentioned above). I have to check, if this point is part of the migration guide..

The other arguments do all boil down to some advances/disadvantages of the underlaying data type. I still think the current default is good as it minimizes surprises (especially for rather inexperienced developers or new users who don't think about the underlaying types). For the more advances users there still is the option to customize this behavior like using a custom converter like mentioned above.

Maybe I could make the default configurable to make it a little bit easier to switch back to ordinal values.

**Comment by Mpdreamz** [comment]
I remember we did this because we got a lot of bug reports where this caught people off guard. 

For sure `[Flag]` enums should be serialised as integers (if they aren't already). 

With https://github.com/elastic/elastic-ingest-dotnet we don't default to writing enums as strings because `System.Text.Json` also doesn't. 

The one argument for sending them as integers is to allow people to rename enums values provided they maintain the enum ordering. 

We should align  https://github.com/elastic/elastic-ingest-dotnet with the client behaviour though.

**Comment by DR9885** [comment]
@Mpdreamz there are also performance benefits to using int over string.
- less network overhead
- less serialization
- faster indexing (text & keyword need larger indexes partial/tokenized matching). Integers only need to be indexed for terms.

I have a few objects where we save with over 100 different enums in each object, with batches of 10k models. The change to will impact performance heavily. 

Also updating each and every field that uses an enum would take a very long time. Can we just make this configurable by the framework?

**Comment by flobernd** [comment]
Changing the new default behavior would be a breaking change at this point.

For folks still searching for a way to be compatible with the original NEST behavior, please have a look at this documentation:
https://www.elastic.co/docs/reference/elasticsearch/clients/dotnet/source-serialization#configuring-custom-jsonserializeroptions

This allows you to remove the globally registered `JsonStringEnumConverter` from the `JsonSerializerOptions.Converter` list.
